### PR TITLE
Port trace_serializer to RapidJSON

### PR DESCRIPTION
**Pull request by dnesbitt61**
# Issue

trace_attributes serialization used legacy JSON code. This PR ports this to use RapidJSON.

A trace_attributes request was used to compare the output on this branch to the output from master (they match):

**http://localhost:8002/trace_attributes?json={"shape":[{"lat": 39.660622, "lon": -76.151993},{"lat": 39.660679, "lon": -76.151817},{"lat": 39.660721, "lon": -76.151711},{"lat": 39.660759, "lon": -76.151588},{"lat": 39.660782, "lon": -76.151421},{"lat": 39.660801, "lon": -76.151268},{"lat": 39.660820, "lon": -76.150810},{"lat": 39.660820, "lon": -76.150681},{"lat": 39.660820, "lon": -76.150452},{"lat": 39.660820, "lon": -76.150070},{"lat": 39.660820, "lon": -76.149872},{"lat": 39.660839, "lon": -76.149673},{"lat": 39.660870, "lon": -76.149559},{"lat": 39.660919, "lon": -76.149460},{"lat": 39.660992, "lon": -76.149330},{"lat": 39.661098, "lon": -76.149147},{"lat": 39.661209, "lon": -76.148972},{"lat": 39.661320, "lon": -76.148781},{"lat": 39.661411, "lon": -76.148651},{"lat": 39.661690, "lon": -76.148262},{"lat": 39.661850, "lon": -76.148003},{"lat": 39.661819, "lon": -76.147888},{"lat": 39.661678, "lon": -76.147469},{"lat": 39.661610, "lon": -76.147263},{"lat": 39.661301, "lon": -76.146378},{"lat": 39.661228, "lon": -76.146172},{"lat": 39.661179, "lon": -76.145714},{"lat": 39.661110, "lon": -76.145279},{"lat": 39.660980, "lon": -76.145020},{"lat": 39.660839, "lon": -76.144920},{"lat": 39.660660, "lon": -76.144791},{"lat": 39.660542, "lon": -76.144691},{"lat": 39.660381, "lon": -76.144623},{"lat": 39.660198, "lon": -76.144600},{"lat": 39.660049, "lon": -76.144592}],"shape_match":"map_snap", "costing":"bicycle", "shape_format": "polyline5"}**

## Tasklist

 - [ ] Add tests
 - [ ] Add #fixes with the issue number that this PR addresses
 - [ ] Update the docs with any new request parameters or changes to behavior described
 - [ ] Update the [changelog](CHANGELOG.md)
 - [ ] If you made changes to the lua files, update the [taginfo](taginfo.json) too.

## Requirements / Relations

 Link any requirements here. Other pull requests this PR is based on?

### CODE REVIEW

**Comment by kevinkreiser** [comment]
nice pr number

**Comment by dnesbitt61** [comment]
will do. My initial claim that this matches master is incorrect. I need to fix this and will format afterwards. Thanks!

File: src/tyr/trace_serializer.cc [hunk 1]
```diff
@@ -510,42 +501,39 @@ std::string serializeTraceAttributes(
...
   bool best_path = true;
-  auto alt_paths_array = json::array({});
-  json->emplace("alternate_paths", alt_paths_array);
   auto route = request.trip().routes().begin();
   for (const auto& map_match_result : map_match_results) {
     if (best_path) {
       // Append the best path trace info
-      append_trace_info(json, controller, request.options(), map_match_result, route->legs(0));
+      append_trace_info(writer, controller, request.options(), map_match_result, route->legs(0));
       best_path = false;
```
Comment (kevinkreiser) [inline src/tyr/trace_serializer.cc:526]:
```suggestion
      best_path = false;
      writer.start_array("alternate_paths");
```

File: src/tyr/trace_serializer.cc [hunk 2]
```diff
@@ -510,42 +501,39 @@ std::string serializeTraceAttributes(
...
     } else {
       // Append alternate path trace info to alternate path array
-      auto alt_path_json = json::map({});
-      append_trace_info(alt_path_json, controller, request.options(), map_match_result,
-                        route->legs(0));
-      alt_paths_array->push_back(alt_path_json);
+      writer.start_array("alternate_paths");
+      append_trace_info(writer, controller, request.options(), map_match_result, route->legs(0));
+      writer.end_array();
```
Comment (kevinkreiser) [inline src/tyr/trace_serializer.cc:531]:
```suggestion
      writer.start_object();
      append_trace_info(writer, controller, request.options(), map_match_result, route->legs(0));
      writer.end_object();
```

File: src/tyr/trace_serializer.cc [hunk 3]
```diff
@@ -510,42 +501,39 @@ std::string serializeTraceAttributes(
...
     }
     ++route;
   }
-  std::stringstream ss;
-  ss << *json;
-  return ss.str();
+  writer.end_object();
```
Comment (kevinkreiser) [inline src/tyr/trace_serializer.cc:535]:
```suggestion
  writer.end_array();
  writer.end_object();
```

**Comment by kevinkreiser** [comment]
@dnesbitt61 is this ready?

**Comment by dnesbitt61** [comment]
yes. test now pass and my comparison of responses before and after look fine (not that I did an exhaustive check).

**Comment by kevinkreiser** [comment]
We still want to merge this because it would be great to get rid of our homegrown json serializer but, since we are on proto3 now all of the `has_` methods are gone. This pr seems to have switched from using the `AttributesController` to just checking if a protobuf field was set (which is probably also a regression anyway). So we should go back to using the controller when serializing then we can merge this.

**Comment by dnesbitt61** [comment]
Merged master into this branch and fixed the use of proto has checks to use AttributesController instead. One question remains on what to set mean elevation to when mean is kNoElevationData.

**Comment by kevinkreiser** [comment]
@dnesbitt61 we can either serialize `null` or just not serialize it. or do you have something else in mind? i don't have strong feelings about it!

**Comment by dnesbitt61** [comment]
This PR was started over a year ago! As such I think it predated the switch to proto3 - that is why it still had all the has_ methods.

**Comment by kevinkreiser** [comment]
@dnesbitt61 yep i know, apologies for that, we should have merged it quickly but its difficult sometimes to keep up with all the activity.

@nilsnolde and i have a bi-weekly meeting where we review these now. we are hoping to burn down the list to something manageable and keep it that way. let us know if you want to join as well, the more the merrier. we put out a call for interest some while ago but there weren't any people who wanted in on the thankless work :wink:

**Review by nilsnolde — DISMISSED** [review]
LGTM

**Review by nilsnolde — APPROVED** [review]

**Pull request merged**